apsuite/commissioning_scripts/meas_coupling_tune.py

```diff
@@ -89,225 +102,144 @@ def _do_meas(self):
+            _plt.savefig(fname, format='png', dpi=300)
         _plt.show()
 
-    def _calc_init_parms(self):
-        """."""
-        parm = self._param
-        vyb = self._tune1[0], self._tune2[0]
-        vye = self._tune1[-1], self._tune2[-1]
-        dparm = parm[-1] - parm[0]
-        coeff1 = (min(vye) - max(vyb)) / dparm
-        offset1 = max(vyb) - coeff1 * parm[0]
-        coeff2 = (max(vye) - min(vyb)) / dparm
-        offset2 = min(vyb) - coeff2 * parm[0]
-        coupling = min(_np.abs(self._tune1 - self._tune2))
-        position = [coeff1, offset1, coeff2, offset2, coupling]
-        self.position = position
+    def _filter_data(self, coup_resolution=None):
+        qcurr = _np.asarray(self.data['current'])
+        tune1, tune2 = self.data['tunes'].T
+        tune1 = _np.asarray(tune1)
+        tune2 = _np.asarray(tune2)
+
+        # sort tune1 > tune2 at each point
+        sel = tune1 <= tune2
+        if sel.any():
+            tune1[sel], tune2[sel] = tune2[sel], tune1[sel]
+
+        if coup_resolution:
+            MeasCoupling.COUPLING_RESOLUTION = coup_resolution
+
+        # remove nearly degenerate measurement points
+        dtunes = _np.abs(tune1 - tune2)
+        sel = _np.where(dtunes < MeasCoupling.COUPLING_RESOLUTION)[0]
+        qcurr = _np.delete(qcurr, sel)
+        tune1 = _np.delete(tune1, sel)
+        tune2 = _np.delete(tune2, sel)
+        return qcurr, tune1, tune2
+
+    def _calc_fitting_error(self):
+        # based on fitting error calculation of scipy.optimization.curve_fit
+        # do Moore-Penrose inverse discarding zero singular values.
+        fit_params = self.analysis['fitted_param']
+        _, smat, vhmat = _np.linalg.svd(
+            fit_params['jac'], full_matrices=False)
+        thre = _np.finfo(float).eps * max(fit_params['jac'].shape)
+        thre *= smat[0]
+        smat = smat[smat > thre]
+        vhmat = vhmat[:smat.size]
+        pcov = _np.dot(vhmat.T / smat*smat, vhmat)
+
+        # multiply covariance matrix by residue 2-norm
+        ysize = len(fit_params['fun'])
+        cost = 2 * fit_params['cost']  # res.cost is half sum of squares!
+        popt = fit_params['x']
+        if ysize > popt.size:
+            # normalized by degrees of freedom
+            s_sq = cost / (ysize - popt.size)
+            pcov = pcov * s_sq
+        else:
+            pcov.fill(_np.nan)
+            print('# of fitting parameters larger than # of data points!')
+        return _np.sqrt(_np.diag(pcov))
+
+    # static methods
+    @staticmethod
+    def _get_normal_modes(params, curr):
+        coeff1, offset1, coeff2, offset2, coupling = params
+        fx_ = coeff1 * curr + offset1
+        fy_ = coeff2 * curr + offset2
+        tune1, tune2 = _np.zeros(curr.shape), _np.zeros(curr.shape)
+        for idx, _ in enumerate(curr):
+            mat = _np.array([[fx_[idx], coupling/2], [coupling/2, fy_[idx]]])
+            eigvalues, _ = _np.linalg.eig(mat)
```

Comment:
Since we are dealing with a real symmetric matrix, I think you can use `eigh` instead of `eig` here. Actually, you can use `eigvalsh`, since we don't care for eigenvectors.

Reply:
Thanks for the advice, now the code uses `eigvalsh`!